Add transformer example (#354)
kubeflow/kfserving: docs/samples/transformer/image_transformer/README.md

# Predict on a KFService using PyTorch Server and Transformer

Most of model servers expect tensors as input data, so a pre-process step is needed before making the prediction call if user is sending in raw input format. Transformer is
a service we orchestrated from KFService spec for user implemented pre/post process code. In the [pytorch](../../pytorch/README.md) example we send the prediction endpoint with

> tensor input in this example we add additional pre-process step to allow user sending in raw image data.

Reviewer: ```suggestion
Most of the model servers expect tensors as input data, so a pre-processing step is needed before making the prediction call if the user is sending in raw input format. Transformer is a service we orchestrated from KFService spec for user implemented pre/post processing code. In the [pytorch](../../pytorch/README.md) example we call the prediction endpoint with tensor inputs, and in this example we add additional pre-processing step to allow the user send raw image data.
```

## Setup
1. Your ~/.kube/config should point to a cluster with [KFServing installed](https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md#deploy-kfserving).
2. Your cluster's Istio Ingress gateway must be network accessible.
3. Your cluster's Istio Egresss gateway must [allow Google Cloud Storage](https://knative.dev/docs/serving/outbound-network-access/)

> ##  Build transformer image

Reviewer: ```suggestion
##  Build Transformer image
```

> ### Extend Transformer and implement pre/postprocess functions

Reviewer: ```suggestion
### Extend Transformer and implement pre/post processing functions
```

```python
from typing import List, Dict
from kfserving.transformer import Transformer
from PIL import Image
import torchvision.transforms as transforms
import logging
import io
import numpy as np
import base64

transform = transforms.Compose(
        [transforms.ToTensor(),
         transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

def image_transform(instance):
    byte_array = base64.b64decode(instance['image_bytes']['b64'])
    image = Image.open(io.BytesIO(byte_array))
    a = np.asarray(image)
    im = Image.fromarray(a)
    res = transform(im)
    logging.info(res)
    return res.tolist()

class ImageTransformer(Transformer):

    def preprocess(self, inputs: Dict) -> Dict:
        return {'instances': [image_transform(instance) for instance in inputs['instances']]}

    def postprocess(self, inputs: List) -> List:
        return inputs
```

> ### Build transformer docker image

Reviewer: ```suggestion
### Build Transformer docker image
```

Reviewer: We should also hyperlink to the Transformer file here in README, and to the Dockerfile as well

This step can be part of your CI/CD pipeline to continuously build the transformer image version. 
```shell
docker build -t yuzisun/image-transformer:latest -f transformer.Dockerfile .
```

## Create the KFService

> 

Reviewer: ```suggestion
Please use the [YAML file](docs/samples/transformer/image_transformer/image_transformer.yaml) to create the KFService, which includes a Transformer and a Predictor
```
